modules/dnn/src/op_inf_engine.cpp

```diff
@@ -718,19 +741,53 @@ Mat infEngineBlobToMat(const InferenceEngine::Blob::Ptr& blob)
     return Mat(size, CV_32F, (void*)blob->buffer());
 }
 
-InfEngineBackendLayer::InfEngineBackendLayer(const InferenceEngine::DataPtr& output_)
+InfEngineBackendLayer::InfEngineBackendLayer(const InferenceEngine::DataPtr& output_,
+                                                   InferenceEngine::CNNNetwork t_net_)
 {
     output = output_;
+    t_net = t_net_;
 }
 
 bool InfEngineBackendLayer::getMemoryShapes(const std::vector<MatShape> &inputs,
                                             const int requiredOutputs,
                                             std::vector<MatShape> &outputs,
                                             std::vector<MatShape> &internals) const
 {
-    std::vector<size_t> dims = output->dims;
-    std::vector<int> shape(dims.rbegin(), dims.rend());
-    outputs.assign(1, shape);
+    InferenceEngine::CNNNetwork curr_t_net(t_net);
+    InferenceEngine::ICNNNetwork::InputShapes inShapes = curr_t_net.getInputShapes();
+    InferenceEngine::ICNNNetwork::InputShapes::iterator itr;
+    bool equal_flag = true;
+    int i = 0;
+    for (itr = inShapes.begin(); itr != inShapes.end(); ++itr)
+    {
+        InferenceEngine::SizeVector currentInShape(inputs[i].begin(), inputs[i].end());
+        if (!std::equal((itr->second).begin(), (itr->second).end(), currentInShape.begin()))
+        {
+            itr->second = currentInShape;
+            equal_flag = false;
+        }
+        i++;
+    }
+    if (!equal_flag)
+    {
+        curr_t_net.reshape(inShapes);
+        InferenceEngine::OutputsDataMap dataOut = curr_t_net.getOutputsInfo();
+        InferenceEngine::OutputsDataMap::iterator itr;
+        std::vector<MatShape> result_out;
+        for (itr = dataOut.begin(); itr != dataOut.end(); ++itr)
+        {
+            std::vector<size_t> dims = itr->second->getDims();
+            MatShape shape(dims.begin(), dims.end());
+            result_out.push_back(shape);
+        }
+        outputs.assign(result_out.begin(), result_out.end());
+    }
+    else
+    {
+        std::vector<size_t> dims = output->dims;
+        std::vector<int> shape(dims.rbegin(), dims.rend());
+        outputs.assign(1, shape);
```

Comment:
Can we remove `output` member completely and extract output shapes from CNNNetwork (calling reshape if needed).